src/redprl/refiner_types.fun

```diff
@@ -0,0 +1,1013 @@
+          T.empty #> (I, H, ty)
+        else
+          raise Fail "Did not match"
+      end
+
+    fun Hyp alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.Hyp"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.VAR (z, O.EXP) = Syn.out tm
+        val CJ.TRUE a = Hyps.lookup H z
+      in
+        T.empty #> (I, H, a)
+      end
+
+    fun Ap alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.Ap"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.AP (m, n) = Syn.out tm
+        val (goalDFun, holeDFun) = makeGoal @@ (I, H) >> CJ.SYNTH m
+        val (goalDom, holeDom) = makeGoal @@ MATCH (O.MONO O.DFUN, 0, holeDFun, [], [])
+        val (goalCod, holeCod) = makeGoal @@ MATCH (O.MONO O.DFUN, 1, holeDFun, [], [n])
+        val (goalN, _) = makeGoal @@ (I, H) >> CJ.MEM (n, holeDom)
+      in
+        T.empty >: goalDFun >: goalDom >: goalCod >: goalN
+          #> (I, H, holeCod)
+      end
+
+    fun S1Elim alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.S1Elim"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.S1_ELIM ((x,cx), m, _) = Syn.out tm
+
+        val cm = substVar (m, x) cx
+        val (goal, _) = makeGoal @@ (I, H) >> CJ.MEM (tm, cm)
+      in
+        T.empty >: goal
+          #> (I, H, cm)
+      end
+
+    fun If alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.If"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.IF ((x,cx), m, _) = Syn.out tm
+
+        val cm = substVar (m, x) cx
+        val (goal, _) = makeGoal @@ (I, H) >> CJ.MEM (tm, cm)
+      in
+        T.empty >: goal
+          #> (I, H, cm)
+      end
+
+    fun PathAp alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.PathAp"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.PATH_AP (m, r) = Syn.out tm
+        val (goalPathTy, holePathTy) = makeGoal @@ (I, H) >> CJ.SYNTH m
+        val (goalLine, holeLine) = makeGoal @@ MATCH (O.MONO O.PATH_TY, 0, holePathTy, [r], [])
+        val psi = T.empty >: goalPathTy >: goalLine
+      in
+        T.empty >: goalPathTy >: goalLine
+          #> (I, H, holeLine)
+      end
+
+    fun Fst alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.Fst"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.FST m = Syn.out tm
+        val (goalTy, holeTy) = makeGoal @@ (I, H) >> CJ.SYNTH m
+        val (goalA, holeA) = makeGoal @@ MATCH (O.MONO O.DPROD, 0, holeTy, [], [])
+      in
+        T.empty >: goalTy >: goalA
+          #> (I, H, holeA)
+      end
+
+    fun Snd alpha jdg =
+      let
+        val _ = RedPrlLog.trace "Synth.Snd"
+        val (I, H) >> CJ.SYNTH tm = jdg
+        val Syn.SND m = Syn.out tm
+        val (goalTy, holeTy) = makeGoal @@ (I, H) >> CJ.SYNTH m
+        val (goalB, holeB) = makeGoal @@ MATCH (O.MONO O.DPROD, 1, holeTy, [], [Syn.into @@ Syn.FST m])
+      in
+        T.empty >: goalTy >: goalB
+          #> (I, H, holeB)
+      end
+  end
+
+  structure Match =
```

Comment:
I think the idea of breaking up the refiner is fine, but this structure (`Match`) probably shouldn't go in this file, since it is not specific to types, but has to do with the general refinement apparatus. Maybe it could go into the main `Refiner` functor or something? Perhaps the same should be done with `Coe` below.